### PR TITLE
DNS server NIO

### PR DESCRIPTION
**Pull request by Mygod**
#2096 but with NIO. Only one thread blocks.

See: https://github.com/shadowsocks/shadowsocks-android/commit/ca25482ff067c4209c19d4b37d51da49fae492f1

Unfortunately there is no NIO SOCKS5 implementation so I wrote one from scratch.

### CODE REVIEW

**Comment by madeye** [comment]
Okay, it looks if there are still on-the-fly DNS queries, shutdown or switching will cause the stuck.

**Comment by Mygod** [comment]
Feel free to investigate lol. I have no idea for now and I will not be looking at this for a while. Also `kotlin-coroutins-debug` does not work on Android for now so this could be painful.

File: core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+                remoteDns.udpUnwrap(result)
+                result
+            }.also { Log.d("forward", "completed $it") }
+        }
+    }
+
+    suspend fun shutdown() {
+        job.cancel()
+        monitor.close()
+        job.join()
```
[the remark this answers is not in the document]
Comment (Mygod) [reply core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: Yeah, we join the job to ensure everything is cleaned up.

**Comment by Mygod** [comment]
@madeye I can't seem to reproduce the issue using your steps...

**Comment by Mygod** [comment]
Anyway check out parallel resolving 96dddba36c2d8748ce547a658399f04b9cc6303f. Yeah IDK.

**Comment by madeye** [comment]
Cool! No more issues on my local device. BTW, it's a Huawei Mate 20.

**Comment by Mygod** [comment]
Could you elaborate more on those steps? I just reproduced it randomly (with your fixed removed ofc lol).

**Comment by Mygod** [comment]
Okay I found the bug. Apparently we need `suspendCancellableCoroutine` instead of `suspendCoroutine`.

**Comment by Mygod** [comment]
Merging this to #2096 as we definitely do not need blocking IO. (unfortunately localResolver is still blocking, hopefully somebody can fix that one day)

File: core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mygod) [reply core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: Okay, I changed my mind. Apparently, there is a good reason not to join this. (see https://github.com/shadowsocks/shadowsocks-android/pull/2096/commits/45a18699ba5595f33c53e576b8a087110b10b67a)

**Pull request closed without merging**